src/convnets/unet.jl

```diff
@@ -0,0 +1,83 @@
+function unet_block(in_chs::Int, out_chs::Int, kernel = (3, 3))
+    return Chain(;
+                 conv1 = Conv(kernel, in_chs => out_chs; pad = (1, 1),
+                              init = _random_normal),
+                 norm1 = BatchNorm(out_chs, relu),
+                 conv2 = Conv(kernel, out_chs => out_chs; pad = (1, 1),
+                              init = _random_normal),
+                 norm2 = BatchNorm(out_chs, relu))
+end
+
+function upconv_block(in_chs::Int, out_chs::Int, kernel = (2, 2))
+    return ConvTranspose(kernel, in_chs => out_chs; stride = (2, 2), init = _random_normal)
+end
+
+"""
+    UNet(inplanes::Integer = 3, outplanes::Integer = 1, init_features::Integer = 32)
+
+    Create a UNet model
+    ([reference](https://arxiv.org/abs/1505.04597v1))
+
+    # Arguments
+    - `in_channels`: The number of input channels
+    - `inplanes`: The number of input planes to the network
+    - `outplanes`: The number of output features
+
+!!! warning
+
+    `UNet` does not currently support pretrained weights.
+"""
+struct UNet
+    encoder::Any
+    decoder::Any
+    upconv::Any
+    pool::Any
+    bottleneck::Any
+    final_conv::Any
+end
+@functor UNet
+
+function UNet(in_channels::Integer = 3, inplanes::Integer = 32,
+              outplanes::Integer = inplanes)
+    features = inplanes
+
+    encoder_layers = []
+    append!(encoder_layers, [unet_block(in_channels, features)])
+    append!(encoder_layers, [unet_block(features * 2^i, features * 2^(i + 1)) for i in 0:2])
+
+    encoder = Chain(encoder_layers)
+
+    decoder = Chain([unet_block(features * 2^(i + 1), features * 2^i) for i in 0:3])
+
+    pool = Chain([MaxPool((2, 2); stride = (2, 2)) for _ in 1:4])
+
+    upconv = Chain([upconv_block(features * 2^(i + 1), features * 2^i) for i in 3:-1:0])
+
+    bottleneck = unet_block(features * 8, features * 16)
+
+    final_conv = Conv((1, 1), features => outplanes)
+
+    return UNet(encoder, decoder, upconv, pool, bottleneck, final_conv)
+end
+
+function (u::UNet)(x::AbstractArray)
+    enc_out = []
+
+    out = x
+    for i in 1:4
+        out = u.encoder[i](out)
+        push!(enc_out, out)
+
+        out = u.pool[i](out)
+    end
+
+    out = u.bottleneck(out)
+
+    for i in 4:-1:1
+        out = u.upconv[5 - i](out)
+        out = cat(out, enc_out[i]; dims = 3)
+        out = u.decoder[i](out)
+    end
+
+    return σ(u.final_conv(out))
+end
```

Comment:
Instead of custom forward pass, can we construct this using `Parallel` from Flux?

Reply:
Either way, the custom forward pass should not use loops as Zygote doesn't do well with them.

Reply:
Either way, the custom forward pass should not use loops as Zygote doesn't do well with them.